### PR TITLE
loop(1077): Redis resilience: withRedis wrapper + convenience helpers + tests in @carbon/kv

### LINKED ISSUE #1077: Redis resilience: wrapper + tests in @carbon/kv (foundation)
## Parent
Part of epic #1076 (Redis downtime kills the entire app).

## Scope
Introduce a resilient accessor layer in `packages/kv/src/client.ts` so all consumers can safely tolerate Redis outages without adding ad-hoc try/catch at each call site.

## What to build

1. **`withRedis<T>(fn: (client) => Promise<T>, fallback: T): Promise<T>` helper** in `packages/kv/src/client.ts`:
   - Wraps any Redis command with a per-call timeout (e.g. 500ms)
   - Catches all connection/command errors (ECONNREFUSED, command timeouts, ioredis offline-queue full)
   - Returns `fallback` on any error — never throws
   - Logs degraded state on first failure + recovery when Redis reconnects (debounced, not per-request)
   - Does not change `enableOfflineQueue` / `retryStrategy` defaults (those govern reconnection, not call safety)

2. **Convenience wrappers:** `safeGet(key) => string | null`, `safeSet(key, value, options?) => void`, `safeDel(key) => void` — all using `withRedis` internally, returning typed defaults on failure.

3. **ioredis-mock unit tests** covering:
   - Normal path: returns real data
   - Redis-down (mock `ioredis` throws): returns fallback, no throw, logs degraded
   - Redis-recovery: subsequent call succeeds after mock is restored

## Acceptance Criteria
- `withRedis` exported from `packages/kv/src/index.ts`; convenience wrappers exported alongside
- No caller can trigger an unhandled rejection via `withRedis`/`safeGet`/`safeSet`/`safeDel`
- Unit tests pass with `ioredis-mock`: normal path, Redis-down fallback, recovery
- TypeScript and biome clean
- No changes to existing consumer call sites in this PR (foundation only)

## Out of scope
Consumer migration (that's #1077-#1080), health endpoint (that's #1081).

### PR DESCRIPTION
**Pull request by carbon-agent**
## Redis resilience: withRedis wrapper + convenience helpers + tests in @carbon/kv

**Kind:** bug · **Risk:** low

Closes #1077

### Acceptance
- [x] withRedis<T>(fn, fallback) exported from packages/kv/src/index.ts; catches all Redis errors (ECONNREFUSED, timeouts, ioredis offline-queue full) and returns fallback without throwing
- [x] safeGet(key) => string | null, safeSet(key, value, options?) => void, safeDel(key) => void exported from packages/kv/src/index.ts; all use withRedis internally
- [x] ioredis-mock unit tests pass: normal path returns real data, Redis-down returns fallback without throw, recovery after mock restored
- [x] No changes to existing consumer call sites (foundation only — no migration in this PR)
- [x] TypeScript typecheck clean (pnpm --filter @carbon/kv tsc --noEmit)
- [x] Biome lint clean (pnpm --filter @carbon/kv run lint or equivalent)

### Tasks
- [x] Add withRedis + safe helpers to client.ts and export from index.ts
- [x] Add ioredis-mock vitest suite for the resilience layer

### Open questions
- Assumption: Assigned Task 1 was already complete/committed and verified clean, so I advanced the same work item by completing Task 2's tests (criterion [2]) this session rather than emitting another no-op checkpoint
- Assumption: Used ioredis-mock (existing devDep) as the controllable in-memory Redis, mocking @carbon/env's REDIS_URL and spying on the singleton's methods to simulate down/slow/recovery; added @types/ioredis-mock (already in the pnpm store) instead of an ambient d.ts since the module resolves to a real untyped JS file that cannot be augmented
- Assumption: Prior committed slice (t1.s5) already added client.test.ts and @types/ioredis-mock; this session confirms all gates green rather than duplicating the file

### Ledger
- 1. **checkpoint** — Added withRedis<T> (500ms timeout + fallback), debounced degraded/recovered logger, and safeGet/safeSet/safeDel to packages/kv/src/client.ts; re-exported all four from index.ts _(slice complete; remaining: Task 2 of 2: add ioredis-mock unit tests (safeGet normal/down, safeSet down, withRedis timeout, recovery) per acceptance criterion [2])_
- 2. **checkpoint** — Verified Task 1 complete: withRedis<T> + debounced logDegraded/logRecovered + safeGet/safeSet/safeDel in packages/kv/src/client.ts, all re-exported from index.ts; typecheck and biome lint clean, no consumer call sites touched _(slice complete; remaining: Task 2 of 2: add ioredis-mock unit tests (safeGet normal/down, safeSet down, withRedis 500ms timeout, recovery) satisfying acceptance criterion [2])_
- 3. **checkpoint** — Verified Task 1 complete — withRedis<T> + debounced logDegraded/logRecovered + safeGet/safeSet/safeDel in packages/kv/src/client.ts, all re-exported from index.ts; typecheck and lint clean, no consumer call sites touched _(slice complete; remaining: Task 2 of 2: add ioredis-mock unit tests satisfying criterion [2] — safeGet normal path returns value, safeGet Redis-down returns null without throw, safeSet Redis-down returns undefined without throw, withRedis 500ms timeout returns fallback, and recovery (throw then succeed returns real data))_
- 4. **checkpoint** — Verified Task 1 complete: withRedis<T> + debounced logDegraded/logRecovered + safeGet/safeSet/safeDel in packages/kv/src/client.ts, all re-exported from index.ts; typecheck (tsgo) and biome lint clean, no consumer call sites touched _(slice complete; remaining: Task 2 of 2: add ioredis-mock unit tests (packages/kv, vitest) satisfying criterion [2] — safeGet normal path returns value, safeGet Redis-down returns null without throw, safeSet Redis-down returns undefined without throw, withRedis 500ms timeout returns fallback, and recovery (throw then succeed returns real data); run `pnpm --filter @carbon/kv test`)_
- 5. **keep** — Added ioredis-mock unit tests for withRedis/safeGet/safeSet/safeDel (normal, Redis-down fallback, 500ms timeout, recovery) in packages/kv/src/client.test.ts; added @types/ioredis-mock devDep so typecheck resolves the mock _(gates green; judge approved)_
- 6. **keep** — Verified Task 2 complete: ioredis-mock vitest suite in packages/kv/src/client.test.ts passes (62 tests), typecheck (tsgo) and biome lint clean; no source edits needed this session — prior slice t1.s5 already landed it _(gates green; judge approved)_

_Conducted headless: 2 kept / 6 iterations. Gated PR — requires human review, do not auto-merge._

## Summary by CodeRabbit

* **New Features**
  * Added resilient Redis support with a safe, timeout-bounded client and new helper APIs (`withRedis`, `safeGet`, `safeSet`, `safeDel`).
  * Default Redis access now degrades gracefully with command-appropriate fallback values and recovery messaging.
* **Bug Fixes**
  * Improved reliability during Redis outages/slowdowns, including safer handling for pipelined/multi operations while preserving blocking command semantics.
* **Tests**
  * Added comprehensive unit tests using an in-memory Redis mock, including timeout, fallback, recovery logging, and pipeline behavior.

### CODE REVIEW

**Comment by barbinbrad** [comment]
@carbon-agent how would this be used? this is a pretty ugly API if i'm reading it right

**Comment by carbon-agent** [comment]
Fair feedback. `withRedis` is the escape hatch for custom operations — but you're right that `withRedis(r => r.get(key), null)` is ugly when you just want a safe get.

The `safeGet` / `safeSet` / `safeDel` helpers are the intended day-to-day API:

```ts
// Before (throws if Redis is down)
const cached = await redis.get(`session:${id}`)

// After (returns null if Redis is down, never throws)
const cached = await safeGet(`session:${id}`)
await safeSet(`session:${id}`, JSON.stringify(data), { ex: 3600 })
await safeDel(`session:${id}`)
```

`withRedis` is only for anything beyond get/set/del (e.g. `INCR`, `ZADD`, multi-key ops). Most callers won't touch it.

That said — if you'd prefer I expand the safe helpers to cover more commands (`safeIncr`, `saddSafe`, etc.) so `withRedis` rarely surfaces, happy to do that. Just say the word.

**Comment by barbinbrad** [comment]
> Fair feedback. `withRedis` is the escape hatch for custom operations — but you're right that `withRedis(r => r.get(key), null)` is ugly when you just want a safe get.
> 
> The `safeGet` / `safeSet` / `safeDel` helpers are the intended day-to-day API:
> 
> ```ts
> // Before (throws if Redis is down)
> const cached = await redis.get(`session:${id}`)
> 
> // After (returns null if Redis is down, never throws)
> const cached = await safeGet(`session:${id}`)
> await safeSet(`session:${id}`, JSON.stringify(data), { ex: 3600 })
> await safeDel(`session:${id}`)
> ```
> 
> `withRedis` is only for anything beyond get/set/del (e.g. `INCR`, `ZADD`, multi-key ops). Most callers won't touch it.
> 
> That said — if you'd prefer I expand the safe helpers to cover more commands (`safeIncr`, `saddSafe`, etc.) so `withRedis` rarely surfaces, happy to do that. Just say the word.

Is it possible to maintain the same API -- but to make it safe without thinking about it?

**Comment by carbon-agent** [comment]
Got it — the goal is zero-migration safety: callers use `redis.get()`, `redis.set()`, etc. exactly as before, but every command silently degrades when Redis is down instead of throwing.

I'll rework the client to wrap the ioredis instance in a `Proxy` that intercepts method calls: each Redis command gets the 500ms timeout + fallback treatment transparently. No `safeGet`/`safeSet`, no API change for consumers — the same `import redis from '@carbon/kv'` just works safely.

Dispatching a fix now.

**Comment by carbon-agent** [comment]
## Rework summary (Proxy-based redesign)

Addressing @barbinbrad's feedback — the original design added explicit \`safeGet\`/\`safeSet\`/\`safeDel\` helpers which required migrating all call sites. The redesign uses a transparent \`Proxy\`:

**Before** (would require migrating every caller):
\`\`\`ts
import redis from '@carbon/kv'
const val = await redis.get('key')  // throws if Redis is down
\`\`\`

**After** (zero migration — same import, same API, silently degrades):
\`\`\`ts
import redis from '@carbon/kv'
const val = await redis.get('key')  // returns null if Redis is down, never throws
\`\`\`

**What changed in the rework commit:**
- Removed \`safeGet\`/\`safeSet\`/\`safeDel\` — no longer needed
- The default \`redis\` export is now a transparent \`Proxy\` (\`createSafeRedis\`) that wraps every command with 500ms timeout + type-appropriate fallback (\`null\` / \`[]\` / \`{}\` / \`0\` / \`"OK"\`) when Redis is unreachable
- \`pipeline()\`/\`multi()\` also wrapped — \`exec()\` returns \`[]\` on failure, chaining works normally on healthy path
- EventEmitter, lifecycle (\`connect\`/\`quit\`/etc.), and stream methods pass through unwrapped
- \`withRedis\` kept as lower-level escape hatch for multi-step custom ops
- Test suite rewritten to exercise the proxy directly
- All CI green on the proxy commit (\`c7f43ec\`)

File: packages/kv/src/client.ts [hunk 1]
```diff
@@ -21,6 +21,301 @@ if (!global.__redis) {
...
+      if (typeof prop !== "string" || PASSTHROUGH.has(prop)) {
+        return (value as AnyFn).bind(target);
+      }
+
+      const command = prop;
+      return (...args: unknown[]) =>
+        runSafe(
+          () => (value as AnyFn).apply(target, args) as Promise<unknown>,
+          fallbackFor(command)
+        );
```
[the remark this answers is not in the document]
Comment (carbon-agent) [reply packages/kv/src/client.ts:330]: Correct — blocking commands with a BLOCK argument are designed to wait indefinitely (or up to their own timeout), so our 500ms wrapper is counterproductive. Will route blpop/brpop/blmove/brpoplpush/bzpopmin/bzpopmax/xread/xreadgroup/WAIT through rawRedis and document the distinction.

File: packages/kv/src/client.ts [hunk 2]
```diff
@@ -21,6 +21,301 @@ if (!global.__redis) {
...
+  "hmget",
+  "hvals",
+  "hkeys",
+  "lrange",
+  "smembers",
+  "sinter",
+  "sunion",
+  "sdiff",
+  "srandmember",
+  "spop",
```
[the remark this answers is not in the document]
Comment (carbon-agent) [reply packages/kv/src/client.ts]: Valid catch. spop and srandmember without a count argument return a single member (string | null), not an array. Will fix the ARRAY_REPLY classification to handle the count-based overload separately.

**Comment by barbinbrad** [comment]
stop working on this issue

**Comment by carbon-agent** [comment]
Understood — stopping work on this. Standing by for further direction.
